File: apps/web/src/test/staffing.test.ts
import { describe, it, expect } from 'vitest';
import { staffingGaps, gapCount, type StaffingMember } from '../logic/staffing';

describe('staffingGaps (#12 leadership/staffing gaps)', () => {
  it('flags filled vs gap required roles, with holder names', () => {
    const roster: StaffingMember[] = [
      { position: 'Ward Mission Leader', person: 'A' },
      { position: 'Ward Missionary', person: 'B' },
      { position: 'Ward Missionary', person: 'C' },
      { position: 'Elders Quorum President', person: 'D' },
      { position: 'Relief Society First Counselor', person: 'E' }, // no RS president
      { position: 'Assistant Ward Mission Leader', person: 'F' }, // not the WML
    ];
    const by = Object.fromEntries(staffingGaps(roster).map((r) => [r.key, r]));
    expect(by.wml.ok).toBe(true);
    expect(by.wml.holders).toEqual(['A']);
    expect(by.ward_missionaries.have).toBe(2);
    expect(by.ward_missionaries.ok).toBe(true);
    expect(by.eq_pres.ok).toBe(true);
    expect(by.rs_pres.ok).toBe(false); // only a counselor, no president
    expect(by.rs_pres.have).toBe(0);
  });

  it('an assistant WML is not the WML; 1 of 2 missionaries is a gap', () => {
    const thin: StaffingMember[] = [
      { position: 'Assistant Ward Mission Leader', person: 'F' },
      { position: 'Ward Missionary', person: 'B' },
    ];
    const by = Object.fromEntries(staffingGaps(thin).map((r) => [r.key, r]));
    expect(by.wml.ok).toBe(false);
    expect(by.ward_missionaries.ok).toBe(false);
    expect(gapCount(thin)).toBe(4); // wml, ward_missionaries, eq, rs all unmet
  });

  it('a fully-staffed unit has zero gaps', () => {
    const full: StaffingMember[] = [
      { position: 'Ward Mission Leader', person: 'A' },
      { position: 'Ward Missionary', person: 'B' },
      { position: 'Ward Missionary', person: 'C' },
      { position: 'Elders Quorum President', person: 'D' },
      { position: 'Relief Society President', person: 'E' },
    ];
    expect(gapCount(full)).toBe(0);
  });
});

File: native/android/app/src/test/java/org/membercovenantpath/viewer/logic/StaffingTest.kt
package org.membercovenantpath.viewer.logic

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Verifies the leadership/staffing gap rule (#12) mirrors web logic/staffing.ts exactly. */
class StaffingTest {

    private fun roster(vararg positions: String) =
        positions.mapIndexed { i, p -> StaffingMember(position = p, person = "Person $i") }

    @Test fun fullyStaffedWardHasNoGaps() {
        val r = roster(
            "Ward Mission Leader",
            "Ward Missionary", "Ward Missionary",
            "Elders Quorum President",
            "Relief Society President",
        )
        assertEquals(0, Staffing.gapCount(r))
        assertTrue(Staffing.gaps(r).all { it.ok })
    }

    @Test fun missingWmlAndTooFewMissionariesAreGaps() {
        // No WML; only 1 ward missionary (need 2); EQ + RS present.
        val r = roster(
            "Ward Missionary",
            "Elders Quorum President",
            "Relief Society President",
        )
        val byKey = Staffing.gaps(r).associateBy { it.key }
        assertFalse(byKey["wml"]!!.ok)              // no ward mission leader
        assertFalse(byKey["ward_missionaries"]!!.ok) // 1 < 2
        assertEquals(1, byKey["ward_missionaries"]!!.have)
        assertTrue(byKey["eq_pres"]!!.ok)
        assertTrue(byKey["rs_pres"]!!.ok)
        assertEquals(2, Staffing.gapCount(r))
    }

    @Test fun assistantWmlDoesNotFillTheLeaderSlot() {
        // An "Assistant Ward Mission Leader" must NOT satisfy the WML requirement.
        val r = roster("Assistant Ward Mission Leader")
        assertFalse(Staffing.gaps(r).first { it.key == "wml" }.ok)
    }

    @Test fun counselorsDoNotFillPresidentSlots() {
        val r = roster(
            "Elders Quorum First Counselor",
            "Relief Society Secretary",
        )
        val byKey = Staffing.gaps(r).associateBy { it.key }
        assertFalse(byKey["eq_pres"]!!.ok) // counselor ≠ president
        assertFalse(byKey["rs_pres"]!!.ok) // secretary ≠ president
    }

    @Test fun branchSynonymsCount() {
        val r = roster(
            "Branch Mission Leader",
            "Branch Missionary", "Branch Missionary",
        )
        val byKey = Staffing.gaps(r).associateBy { it.key }
        assertTrue(byKey["wml"]!!.ok)
        assertTrue(byKey["ward_missionaries"]!!.ok)
    }

    @Test fun holdersListNamesTheServingPeople() {
        val r = listOf(
            StaffingMember(position = "Ward Mission Leader", person = "Jane Doe"),
            StaffingMember(position = "Ward Missionary", person = "  "), // blank → not a holder
        )
        val wml = Staffing.gaps(r).first { it.key == "wml" }
        assertEquals(listOf("Jane Doe"), wml.holders)
        val wm = Staffing.gaps(r).first { it.key == "ward_missionaries" }
        assertEquals(1, wm.have)            // counted as held…
        assertTrue(wm.holders.isEmpty())    // …but the blank name is not listed
    }
}
